pasture-tools/src/bin/render/renderer.rs

```diff
@@ -0,0 +1,689 @@
+use pasture_core::gpu as pgpu;
+use pasture_io::base::PointReader;
+use pasture_io::las::LASReader;
+
+use pasture_core::{
+    containers::{
+        InterleavedVecPointStorage,
+        PointBufferExt,
+    },
+    layout::{
+        attributes, PointLayout, PointAttributeDefinition, PointAttributeDataType
+    },
+};
+
+use winit::window::Window;
+use nalgebra::{Vector2, Vector3, Point3, UnitQuaternion, Matrix4};
+use crevice::std140::AsStd140;
+use instant::Instant;
+
+#[repr(C)]
+#[derive(Debug, AsStd140)]
+struct UboData {
+    view_proj_matrix: mint::ColumnMatrix4<f32>,
+}
+
+pub struct Camera {
+    // camera position and orientation in world space
+    pos: Point3<f32>,
+    rot: UnitQuaternion<f32>,
+
+    // perspective projection parameters
+    aspect: f32,
+    fovy: f32,
+}
+
+impl Camera {
+    pub fn new() -> Camera {
+        Camera {
+            pos: Point3::new(0.0, 0.0, 0.0),
+            rot: UnitQuaternion::identity(),
+            aspect: 1.0,
+            fovy: 1.5f32, // 90 degrees
+        }
+    }
+
+    pub fn view_dir(&self) -> Vector3::<f32> {
+        self.rot * Vector3::new(0.0f32, 0.0f32, -1.0f32)
+    }
+
+    pub fn view_mat(&self) -> nalgebra::Matrix4::<f32> {
+        let up = Vector3::new(0.0f32, 1.0f32, 0.0f32);
+        let target = self.pos + self.view_dir();
+
+        nalgebra::Matrix4::look_at_rh(&self.pos, &target, &up)
+    }
+}
+
+pub trait CameraController {
+    fn update(&mut self, _cam: &mut Camera, _dt: f32) {}
+    fn process_mouse_move(&mut self, _cam: &mut Camera, _pos: Vector2::<f32>) {}
+    fn process_keyboard_input(&mut self, _cam: &mut Camera, _key: winit::event::VirtualKeyCode, _pressed: bool) {}
+    fn process_mouse_button(&mut self, _cam: &mut Camera, _key: winit::event::MouseButton, _pressed: bool) {}
+    fn process_mouse_wheel(&mut self, _cam: &mut Camera, _delta: f32) {}
+}
+
+pub struct FPSCameraController {
+    pitch: f32,
+    yaw: f32,
+
+    // controller information
+    pub w_down: bool,
+    pub a_down: bool,
+    pub s_down: bool,
+    pub d_down: bool,
+    pub q_down: bool,
+    pub e_down: bool,
+    pub rotating: bool,
+    pub mouse_pos: Option<Vector2::<f32>>,
+}
+
+pub struct ArcballCameraController {
+	// Describes how far the rotation center is in front of the camera.
+	offset: f32,
+
+	rotating: bool,
+	panning: bool,
+    mouse_pos: Option<Vector2::<f32>>,
+}
+
+// TODO: we only need a Vec here because gpu_point_buffer needs it
+const POINT_ATTRIB_3D_F32: PointAttributeDefinition = PointAttributeDefinition::custom(
```

Comment:
The current pasture version should support `POSITION_3D.with_custom_datatype(PointAttributeDataType::Vec3f32)` in a constant expression. This is safer than specifying the name `"Position3D"` by hand, as this might change in the future